### DISCUSSION TITLE
Dropdown gets triggered even if value does not change

### DISCUSSION THREAD

**Discussion by gaborfodor** [h2oai/wave#680] [Q&A]
#### Wave SDK Version, OS
0.11.0 Mac
(e.g. 4.2.0, Windows)

#### Actual behavior

I have a single button and a dropdown menu. The dropdown menu gets triggered too when I press the button.

#### Expected behavior
According to doc I expected it to get triggered only when the value changes.
"trigger: True if the form should be submitted when the dropdown value changes."

#### Steps To Reproduce
```
from h2o_wave import ui, Q, app, main

_ = main


@app('/')
async def my_app(q: Q):
    print('=' * 10)
    print(q.args)
    await main_page(q)
    await q.page.save()


async def main_page(q: Q):
    if not q.client.initialized:
        q.client.initialized = True
        q.client.abc = 'a'

    if q.args.abc:
        print('ABC triggered')
        q.client.abc = q.args.abc
    if q.args.T:
        print('True')
    if q.args.F:
        print('False')

    q.page['dropdown'] = ui.form_card(
        box='1 1 1 1', items=[
            ui.dropdown(name='abc', value=q.client.abc, trigger=True, choices=[
                ui.choice(name='a', label='A'),
                ui.choice(name='b', label='B'),
            ])
        ]
    )
    q.page['button'] = ui.form_card(
        box='2 1 2 1', items=[
            ui.buttons([
                ui.button(name='T', label='True', primary=True),
                ui.button(name='F', label='False'),
            ])
        ]

    )

**Comment by lo5** [comment]
@gaborfodor This is by design. All the components on a page submit their values, independent of the dropdown's `trigger`.

Normally, the buttons on a page are the canonical way to send information from the browser to the app. But in some cases, you might want to get that information as soon as the value of a component changes (saves the user from the additional step of having to click some button) - hence the `trigger` facility on several components like the dropdown. 

> According to doc I expected it to get triggered only when the value changes.

It's simply a flag for "should information be submitted when the value changes?" - and not "only when the value changes".

**Comment by gaborfodor** [comment]
Thanks, that makes sense. 
The following solves my issue to trigger something only if the value is changed.  
```
if q.args.abc and q.client.abc != q.args.abc:
    print('ABC triggered')
    q.client.abc = q.args.abc
```

**Comment by goldentom42** [comment]
I've just hit this bug in a funny way

https://user-images.githubusercontent.com/17214417/106151242-71527380-617c-11eb-98db-861e2c8650f6.mov

I'm afraid @gaborfodor 's trick would not work for me, unless I set the dropdown value to None in q.client.
I may do that actually.

**Comment by goldentom42** [comment]
This means we need to ensure there is a hierarchy on the way `q.args` are treated to route user requests.

**Comment by goldentom42** [comment]
I understand the design though :)
